cirq-rigetti/cirq_rigetti/aspen_device.py

```diff
@@ -0,0 +1,514 @@
+        topology.
+
+        Additionally, this method will not attempt to decompose the operation into this
+        device's native gate set. This integration, by default, uses the Quil
+        compiler to do so.
+
+        Please see the Quil Compiler
+        [documentation](https://pyquil-docs.rigetti.com/en/stable/compiler.html)
+        for more information.
+
+        Args:
+            operation: The operation to validate.
+
+        Raises:
+            UnsupportedRigettiQCSOperation: The operation isn't valid for this device.
+        """
+        qubits = operation.qubits
+        for qubit in qubits:
+            self.validate_qubit(qubit)
+        if len(qubits) == 2:
+            i = self._aspen_qubit_index(qubits[0])
+            j = self._aspen_qubit_index(qubits[1])
+            if j not in self.qubit_topology[i]:
+                raise UnsupportedRigettiQCSOperation(
+                    f'qubits {qubits[0]} and {qubits[1]} do not share an edge',
+                )
+
+    def _value_equality_values_(self):
+        return self._maximum_qubit_number
+
+    def __repr__(self):  # pragma: no cover
+        return f'cirq_rigetti.RigettiQCSAspenDevice(isa={self.isa!r})'
+
+    def _json_dict_(self):
+        return {
+            'cirq_type': 'RigettiQCSAspenDevice',
+            'isa': self.isa.to_dict(),
+        }
+
+    @classmethod
+    def _from_json_dict_(cls, isa, **kwargs):
+        return cls(
+            isa=InstructionSetArchitecture.from_dict(isa),
+        )
+
+
+@_provide_default_client
+def get_rigetti_qcs_aspen_device(
+    quantum_processor_id: str,
+    client: Optional[httpx.Client],
+) -> RigettiQCSAspenDevice:
+    """retrieves a `qcs_api_client.models.InstructionSetArchitecture` from the Rigetti
+    QCS API and uses it to initialize a RigettiQCSAspenDevice.
+
+    Args:
+        quantum_processor_id: The identifier of the Rigetti QCS quantum processor.
+        client: Optional; A `httpx.Client` initialized with Rigetti QCS credentials
+        and configuration. If not provided, `qcs_api_client` will initialize a
+        configured client based on configured values in the current user's
+        `~/.qcs` directory or default values.
+
+    Returns:
+        A `RigettiQCSAspenDevice` with the specified quantum processor instruction
+        set and architecture.
+
+    """
+    # coverage: ignore
+    isa = cast(
+        InstructionSetArchitecture,
+        get_instruction_set_architecture(
+            client=client, quantum_processor_id=quantum_processor_id
+        ).parsed,
+    )
+    return RigettiQCSAspenDevice(isa=isa)
+
+
+class OctagonalQubit(cirq.ops.Qid):
+    """A cirq.Qid supporting Octagonal indexing."""
+
+    def __init__(self, octagon_position: int):
+        """Initializes an `OctagonalQubit` using indices 0-7.
+              4  - 3
+            /        \
+          5           2
+          |           |
+          6           1
+            \\       /
```

Comment:
```suggestion
            \        /
```
Did you mean to put two edges between 6 and 7?

Reply:
No, I was [escaping backslash](https://docs.python.org/2.0/ref/strings.html). Probably more clear if I use a raw string.